### PR TITLE
Promote DocBlock to native types

### PR DESCRIPTION
**Pull request by Slamdunk**

### CODE REVIEW

File: src/Annotation/BuilderAbstractFactory.php [hunk 1]
```diff
@@ -73,7 +71,7 @@ public function canCreate(ContainerInterface $container, $requestedName)
      *
      * @return array
```
Comment (Slamdunk) [inline src/Annotation/BuilderAbstractFactory.php:72]: PHP Code Sniffer keeps `@return array` notation to let be more specific if the developer is willing to be so

**Comment by froschdesign** [comment]
@Slamdunk 
This is a big step. I think it would be good if we could add something to the migration guide to help the user, for example using Reactor with some rules for rewriting custom classes.

What do you think?

**Comment by Slamdunk** [comment]
Even though it seems big, not even e single LOC of https://docs.laminas.dev/laminas-form/ needs to be changed (AFAICT).

The only and most common extension point is the `Form`, which typically only gets `init()` and `getInputFilterSpecification()` implemented, none of which requires a change.

Users so advanced to write their own `Elements` or `ViewHelpers` just need to promote the types already in use, for which I can suggest the command I've used if you wish:
```console
$ php-cs-fixer fix --verbose --rules=phpdoc_to_param_type,phpdoc_to_return_type --allow-risky=yes
```

**Comment by froschdesign** [comment]
@Slamdunk 
I mean, a big step in a good and right direction! 😃 

> Users so advanced to write their own `Elements` or `ViewHelpers` just need to promote the types already in use, for which I can suggest the command I've used if you wish:

That would be great! 👍

**Comment by Slamdunk** [comment]
I don't see an UPGRADE file or page, is there a standard to write and publish it in `laminas` org?

**Comment by driehle** [comment]
I created a `migration/v2-to-v3.md` in docs with #123, in accordance to other components, e.g. https://docs.laminas.dev/laminas-router/migration/v2-to-v3/.

**Comment by froschdesign** [comment]
The documentation is the right place for this and a new page is on the way with #123 

https://github.com/laminas/laminas-form/blob/17622276cefbe3f32a64d7623c3d0bc68a8e5b58/docs/book/migration/v2-to-v3.md

File: docs/book/migration/v2-to-v3.md [hunk 2]
```diff
@@ -4,15 +4,27 @@ laminas-form v3 makes a number of changes that may affect your application. This
...
+laminas-form v3 promoted `@param` and `@return` annotations of non-mixed types to
+the corresponding function signatures. This change should affect only highly customized
+element, fieldset or form classes.
+If you need to automate native types promotion, [PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer)
+can help you with this command:
+
+```console
+$ php-cs-fixer fix --rules=phpdoc_to_param_type,phpdoc_to_return_type --allow-risky=yes path/to/my/custom/forms/
+```
+
```
Comment (froschdesign) [inline docs/book/migration/v2-to-v3.md:18]: @Slamdunk 
Thank you! 😃

**Comment by Slamdunk** [comment]
@driehle @froschdesign I'd like to release `3.0.0` after https://github.com/laminas/laminas-form/pull/98#issuecomment-850383139 

I've already tested `3.0.x-dev` in my app and found few bugs like #133, #135 and #136 linked to this PR: could you please test `3.0.x-dev` too in the upcoming days?

**Comment by driehle** [comment]
I am wondering if we should probably consider #62 and #71 before the release of 3.0.0?

**Comment by froschdesign** [comment]
@driehle 
There is still some time before we can release version 3.0.0, because I have to extend the documentation. The versioning is missing and some other updates.

Help is welcome at any time! 😃

**Comment by Slamdunk** [comment]
@driehle since #62 and #71 can be addressed in a minor, this morning I've released `3.0.0` to speed the things up.

Thank you very much for the big help you gave us releasing this important milestone :+1:

**Pull request merged**